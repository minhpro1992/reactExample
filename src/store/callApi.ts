import axios, { AxiosRequestConfig, Method } from 'axios'
// import { ObjectKeyString } from './types'
export type ObjectKeyString<T> = { [key: string]: T }
export async function callApi(
  method: string,
  url: string,
  path: string,
  data?: Record<string, unknown>
) {
  const res = await fetch(`${url}${path}`, {
    method,
    headers: {
      Accept: 'application/json'
    },

    body: JSON.stringify(data)
  })

  return res.json()
}
export const urlHasQuery = (url: string): boolean => {
  // URL already a query if '?' is in URL
  return url.indexOf('?') !== -1
}

export const addQueryToURL = (
  url: string,
  queryParams: ObjectKeyString<string>
): string => {
  let urlWithQuery = url

  Object.keys(queryParams).forEach((key) => {
    // Add query initial '?', if already present add '&'
    urlWithQuery += urlHasQuery(urlWithQuery) ? '&' : '?'
    // Add query param [key=value] to url
    urlWithQuery += `${key}=${queryParams[key] ? queryParams[key] : ''}`
  })

  return urlWithQuery
}

export async function axiosCallApi(
  method: Method,
  url: string,
  path: string,
  data?: Record<string, unknown>,
  headers = 'application/json',
  customHeaders?: Record<string, string>
) {
  const config: AxiosRequestConfig = {
    method: method,
    url: url + path,
    data: data,
    headers: {
      'Content-Type': headers,
      ...customHeaders
    }
  }

  return axios(config).then(
    (response) => response,
    (error) => Promise.reject(error.response)
  )
}
