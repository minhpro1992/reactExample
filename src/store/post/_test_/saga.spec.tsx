import postSaga from "../sagas";
import {
  DELETE_POST_FAIL,
  DELETE_POST_REQUEST,
  DELETE_POST_SUCCESS,
  GET_POST_FAIL,
  GET_POST_REQUEST,
  GET_POST_SUCCESS,
  PostItemType,
} from "../types";
import { expectSaga } from "redux-saga-test-plan";
import * as matchers from "redux-saga-test-plan/matchers";
import { throwError } from "redux-saga-test-plan/providers";
import { callApi } from "../../callApi";

const postStoreState = {
  posts: [],
  isLoading: false,
  error: {},
};
const postId = 1;
const expectedError = new Error("error");
describe("post saga", () => {
  it("handle get post success", async () => {
    const response: PostItemType[] = [
      {
        id: 1,
        userId: 1,
        title: "test",
        body: "test 12345",
      },
    ];
    const result = await expectSaga(postSaga)
      .withState(postStoreState)
      .provide({
        call() {
          return response;
        },
      })
      .put.actionType(GET_POST_SUCCESS)
      .dispatch({ type: GET_POST_REQUEST })
      .silentRun();
    expect(result.toJSON()).toMatchSnapshot();
  });
  it("handle get post fail", () => {
    try {
      return expectSaga(postSaga)
        .provide([[matchers.call.fn(callApi), throwError(expectedError)]])
        .put.actionType(GET_POST_FAIL)
        .dispatch({
          type: GET_POST_REQUEST,
        })
        .silentRun();
    } catch (error) {
      expect(error).toEqual(expectedError);
    }
  });
  it("delete post successfully", async () => {
    const result = await expectSaga(postSaga)
      .withState(postStoreState)
      .provide({
        call() {
          return postId;
        },
      })
      .put.actionType(DELETE_POST_SUCCESS)
      .dispatch({ type: DELETE_POST_REQUEST, payload: postId })
      .silentRun();
    expect(result.toJSON()).toMatchSnapshot();
  });
  it("delete post fail", () => {
    try {
      return expectSaga(postSaga)
        .provide([[matchers.call.fn(callApi), throwError(expectedError)]])
        .put.actionType(DELETE_POST_FAIL)
        .dispatch({ type: DELETE_POST_REQUEST, payload: postId })
        .silentRun();
    } catch (error) {
      expect(error).toEqual(expectedError);
    }
  });
});
