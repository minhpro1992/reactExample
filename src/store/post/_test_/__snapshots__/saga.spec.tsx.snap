// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post saga delete post successfully 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "delete",
          "https://jsonplaceholder.typicode.com",
          "/posts/1",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/callApi",
      },
      "type": "CALL",
    },
  ],
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/watchPostRequest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/watchDeletePostRequest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "GET POST REQUEST",
          "@@redux-saga-test-plan/json/function/handlePost",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "DELETE POST REQUEST",
          "@@redux-saga-test-plan/json/function/handleDeletePost",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "payload": 1,
            "type": "DELETE POST REQUEST",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/handleDeletePost",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "GET POST REQUEST",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "DELETE POST REQUEST",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "DELETE POST REQUEST",
      },
      "type": "TAKE",
    },
  ],
}
`;

exports[`post saga get  post info success 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "delete",
          "https://jsonplaceholder.typicode.com",
          "/posts/1",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/callApi",
      },
      "type": "CALL",
    },
  ],
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/watchPostRequest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/watchDeletePostRequest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "GET POST REQUEST",
          "@@redux-saga-test-plan/json/function/handlePost",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "DELETE POST REQUEST",
          "@@redux-saga-test-plan/json/function/handleDeletePost",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "payload": 1,
            "type": "DELETE POST REQUEST",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/handleDeletePost",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "GET POST REQUEST",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "DELETE POST REQUEST",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "DELETE POST REQUEST",
      },
      "type": "TAKE",
    },
  ],
}
`;

exports[`post saga handle get post success 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "get",
          "https://jsonplaceholder.typicode.com",
          "/posts",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/callApi",
      },
      "type": "CALL",
    },
  ],
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/watchPostRequest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/watchDeletePostRequest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "GET POST REQUEST",
          "@@redux-saga-test-plan/json/function/handlePost",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "DELETE POST REQUEST",
          "@@redux-saga-test-plan/json/function/handleDeletePost",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "type": "GET POST REQUEST",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/handlePost",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "GET POST REQUEST",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "DELETE POST REQUEST",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "GET POST REQUEST",
      },
      "type": "TAKE",
    },
  ],
}
`;
